nflow-engine/src/main/java/com/nitorcreations/nflow/engine/internal/executor/WorkflowStateProcessor.java

```diff
@@ -239,63 +251,155 @@ private WorkflowActionType getActionType(StateExecutionImpl execution) {
-      StateExecutionImpl execution, WorkflowState currentState) {
-    WorkflowStateMethod method = definition.getMethod(instance.state);
-    if (method == null) {
-      execution.setNextState(currentState);
-      return stopInState(currentState, "Execution finished.");
+  private NextAction processWithListeners(ListenerContext listenerContext, final WorkflowInstance instance, final AbstractWorkflowDefinition<? extends WorkflowState> definition,
+                                          final StateExecutionImpl execution, final WorkflowState state) {
+    List<WorkflowExecutorListener> chain = new LinkedList<>(asList(this.executorListeners));
+    ProcessingExecutorListener processingListener = new ProcessingExecutorListener(instance, definition, execution, state);
+    chain.add(processingListener);
+    NextAction nextAction = new ExecutorListenerChain(chain).next(listenerContext);
+    if(execution.isStateProcessed()) {
+      return nextAction;
     }
-    NextAction nextAction;
-    Object[] args = objectMapper.createArguments(execution, method);
-    if (currentState.getType().isFinal()) {
-      invokeMethod(method.method, definition, args);
-      nextAction = stopInState(currentState, "Stopped in final state");
-    } else {
-      try {
-        nextAction = (NextAction) invokeMethod(method.method, definition, args);
-        if (nextAction == null) {
-          logger.error("State '{}' handler method returned null, proceeding to error state '{}'", instance.state, definition
-              .getErrorState().name());
-          nextAction = moveToState(definition.getErrorState(), "State handler method returned null");
-          execution.setFailed();
-        } else if (nextAction.getNextState() != null && !definition.getStates().contains(nextAction.getNextState())) {
-          logger.error("State '{}' is not a state of '{}' workflow definition, proceeding to error state '{}'",
-              nextAction.getNextState(), definition.getType(), definition.getErrorState().name());
-          nextAction = moveToState(definition.getErrorState(), "State '" + instance.state
-              + "' handler method returned invalid next state '" + nextAction.getNextState() + "'");
-          execution.setFailed();
-        } else if (!"ignore".equals(illegalStateChangeAction) && !definition.isAllowedNextAction(instance, nextAction)) {
-          logger.warn("State transition from '{}' to '{}' is not allowed by workflow definition.", instance.state,
-              nextAction.getNextState());
-          if ("fail".equals(illegalStateChangeAction)) {
-            nextAction = moveToState(definition.getErrorState(), "Illegal state transition from " + instance.state + " to "
-                + nextAction.getNextState().name() + ", proceeding to error state " + definition.getErrorState().name());
+    return new SkippedStateHandler(nextAction, instance, definition, execution, state, illegalStateChangeAction, objectMapper).processState();
```

Comment:
avoid passing class fields as parameters to static inner classes, make them package private instead